ImfInput now rounds sub-millisecond times on write.

spacecataz opened the pull request:
Closes #291 

Repeat of previous PR that had conflicting merges.

Changes writing ImfInput objects to file: sub-millisecond values now use floor when converting micro- to millisecond values instead of round.
A new test has been added to make sure that this works using the example given in the issue.

I want to note @jtniehof's suggestion:

```
t = self.time[i]
ms = int(round(t.microsecond / 1000.))
if ms = 10000:
    ms = 0
    t += dt.timedelta(microseconds=1000)
out.write('{:%Y %m %d %H %M %S} {:03d} '.format(t, ms))
```

This is an alternative worth discussing, but I am hesitant to nest an `if` statement in a sometimes very long `for` loop.  IMF files can contain upwards of month(s) worth of data at sub-minute resolution, so writing can already be slow.  Worth discussing.

PR Checklist
[ n/a] New code has inline comments where necessary
[ n/a] Any new modules, functions or classes have docstrings consistent with SpacePy style
[ n/a] Added an entry to CHANGELOG if fixing a major bug or providing a major new feature
[ x] New features and bug fixes should have unit tests

drsteve commented on tests/test_pybats.py:
```diff
@@ -329,12 +329,32 @@ def setUp(self):
+
+        # Write and test for non-failure on re-read:
+        imf.write('testme.tmp')
+        imf2 = pb.ImfInput('testme.tmp')
```
This doesn't get cleaned up... can we use a temporary file or have some code to remove the file after the test? (I assume that if, for some reason, the write fails then there's nothing to clean up)

spacecataz commented on tests/test_pybats.py:
```diff
@@ -329,12 +329,32 @@ def setUp(self):
     knownImfIono = [4.99, 0.01]
+    knownSubMilli= dt.datetime(2017, 9, 6, 16, 42, 36, 999000)
 
     def tearDown(self):
```
Can confirm that tearDown exists and works properly on my machine.  @drsteve, can you confirm that tearDown is working for you?

jtniehof commented:
You're doing a lot more expensive stuff in that loop than an if. Unless
there's some evidence that's a substantial contributor, I think it makes
more sense to round. I can do a dumb test but if you've got a sample big
file that would be better.

On Mon, May 4, 2020, 19:16 spacecataz <notifications@github.com> wrote:

> Closes #291 <https://github.com/spacepy/spacepy/issues/291>
>
> Repeat of previous PR that had conflicting merges.
>
> Changes writing ImfInput objects to file: sub-millisecond values now use
> floor when converting micro- to millisecond values instead of round.
> A new test has been added to make sure that this works using the example
> given in the issue.
>
> I want to note @jtniehof <https://github.com/jtniehof>'s suggestion:
>
> t = self.time[i]
> ms = int(round(t.microsecond / 1000.))
> if ms = 10000:
>     ms = 0
>     t += dt.timedelta(microseconds=1000)
> out.write('{:%Y %m %d %H %M %S} {:03d} '.format(t, ms))
>
> This is an alternative worth discussing, but I am hesitant to nest an if
> statement in a sometimes very long for loop. IMF files can contain
> upwards of month(s) worth of data at sub-minute resolution, so writing can
> already be slow. Worth discussing.
>
> PR Checklist
> [ n/a] New code has inline comments where necessary
> [ n/a] Any new modules, functions or classes have docstrings consistent
> with SpacePy style
> [ n/a] Added an entry to CHANGELOG if fixing a major bug or providing a
> major new feature
> [ x] New features and bug fixes should have unit tests
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/spacepy/spacepy/pull/298
> Commit Summary
>
>    - ImfInput now rounds down sub-millisecond times on write.
>
> File Changes
>
>    - *M* spacepy/pybats/__init__.py
>    <https://github.com/spacepy/spacepy/pull/298/files#diff-5ca70361a60eb7f076b7789d7c1ee733>
>    (2)
>    - *M* tests/test_pybats.py
>    <https://github.com/spacepy/spacepy/pull/298/files#diff-524c1b470567b0c24086f715b1fcdee7>
>    (20)
>
> Patch Links:
>
>    - https://github.com/spacepy/spacepy/pull/298.patch
>    - https://github.com/spacepy/spacepy/pull/298.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/spacepy/spacepy/pull/298>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABP5W4VIZE6J6B5JB3DX6DRP5EFJANCNFSM4MZE7TRQ>
> .
>

drsteve commented:
I can provide a chunky input file... Let me just find it and I can send a link.

spacecataz commented:
I used the file provided by @drsteve, got average write times over 50 writes, and compared with the extra rounding work and without.  Minor impact, so I've added that change.  Have also updated the test.  I think we're good to go here.
